src/common/directed_graph.h

```diff
@@ -0,0 +1,201 @@
+  /*
```

Comment:
Is this PR trying to address: #15198 . My understanding was that this would be something on the lines of set_monitor_callback for gluon to monitor intermediate nodes in the graph. This seems to be addressing a different problem of IR visualization. For the forward pass, the SaveJSON coupled with visualization already seems to be doing the trick.(which can be extended for backward too if needed) Also , I see TVM has a PrintGraphIR nnvm pass that we can leverage. Do we really need this ?

Reply:
Good question, the intention is only to dump the graph in Backward to dot file or json, so to be precise is only tangential to the original issue. I haven't used set_monitor_callback, but from a quick look seems it's bound to an executor. How would you use it for backward in imperative? I'm not sure where to get an executor in that case.

where is the SaveJSON pass defined? I'm having trouble finding it now due to the lack of types: 

https://github.com/dmlc/tvm/blob/21935dcbf56ad3bd66ebff9891a6bc3865b8106d/nnvm/include/nnvm/pass_functions.h#L59

Could you point to the implementation of SaveJSON pass and the visualization code? Does it convert to dot?

Reply:
https://github.com/dmlc/tvm/blob/21935dcbf56ad3bd66ebff9891a6bc3865b8106d/nnvm/src/pass/saveload_json.cc#L314 . You can find visualization code here, It uses digraph to save it in dot format: https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/visualization.py

Reply:
> Good question, the intention is only to dump the graph in Backward to dot file or json, so to be precise is only tangential to the original issue. I haven't used set_monitor_callback, but from a quick look seems it's bound to an executor. How would you use it for backward in imperative? I'm not sure where to get an executor in that case.

How to add something set_monitor_callback would require a deepdive but I think it would be somewhere in the cached_op. Another question is in gluon would you want to monitor the tensors at the block level or the symbol level. That also will change things.

Reply:
As discussed today, I will only dump the graph and not tensor values. So dumping tensor values it out of scope for this PR. So I won't deal with monitor callback.

Reply:
can you summarize why we need this ? if its to visualize the backward pass, why cant we extend the tools that we have ?

Reply:
I thought a bit about this and I think you have a valid point, while we could serialize to json using the pass that you mention. Is not useful to call python code to visualize backward, since we want this from inside the engine. So I think for us now makes sense to have a couple of files with the logic to do the custom dump how we want.

I think is desirable to dump to DOT and not only JSON. Also the Symbol2JSONGraph that ends up being called from the links you provided is in the tvm subrepo and I think is best to have the code for the dump in the mxnet repo for modifications.

Reply:
I guess then, the purpose of the PR is to add visualization tool which would directly dump to dot format instead of using the frontend tool to do a followup pass on the symbol json. I think adding so much more code to avoid one line is a bit too much. I am not blocking this though since we are not exposing this to customer but I think it is not needed.

Reply:
What do you suggest exactly? using directly Symbol2JSONGraph?

Reply:
I'm just adding a generic graph class with unit tests. I think it might be useful to have as it's generic.

Reply:
Create a dummy graph with same outputs as the graph you want to visualize and call SaveJSON on it. Visualize it with python visualization tool.

Reply:
I don't this this is practical as it uses symbol plus needs additional python. I think in this case my code does the job in pure C++ without any needs for python. After considering your suggestion I think it's best to use this C++ approach. The graph is generic, reusable and tested so I don't see a big problem with this.